src/failure_message.rs

```diff
@@ -76,40 +101,114 @@ pub struct MessagePayload {
     failures: Vec<Failure>,
 }
 
+#[derive(Serialize)]
+struct TemplateContext {
+    num_tests: i32,
+    num_failed: i32,
+    num_passed: i32,
+    num_skipped: i32,
+    num_output: i32,
+    failures: Vec<TemplateFailure>,
+}
+
+impl TemplateContext {
+    fn new(
+        num_tests: i32,
+        num_failed: i32,
+        num_passed: i32,
+        num_skipped: i32,
+        failures: Vec<TemplateFailure>,
+    ) -> Self {
+        let num_output: i32 = failures.len().try_into().unwrap();
+        Self { num_tests, num_failed, num_passed, num_skipped, num_output, failures }
+    }
+}
+
+#[derive(Serialize)]
+struct TemplateFailure {
+    test_suite: String,
+    test_name: String,
+    duration: String,
+    backticks: String,
+    build_url: Option<String>,
+    stack_trace: Vec<String>,
+}
+
+impl TemplateFailure {
+    fn new(
+        test_suite: String, 
+        test_name: String, 
+        duration: String, 
+        raw_num_backticks: usize, 
+        build_url: Option<String>,
+        stack_trace: Vec<String>
+    ) -> Self {
+        let num_backticks = max(raw_num_backticks + 1, 3);
+        let backticks = String::from("`".repeat(num_backticks));
+        Self { test_suite, test_name, duration, backticks, build_url, stack_trace }
+    }
+}
+
+fn longest_repeated_substring(s: String, target: char) -> usize {
```

Comment:
```suggestion
fn longest_repeated_substring(s: &str, target: char) -> usize {
```

Unless you actually want to *move ownership*, using `&str` is preferred over `String`.